### ISSUE TITLE
Implement configuration specific imports

### ISSUE THREAD

**Issue by floitschG** [dart-lang/sdk#24581]
Implement [configuration specific imports](https://github.com/munificent/dep-interface-libraries/blob/master/Proposal.md) [0] behind a flag.
Validate it, and then remove the flag, once the DEP committee has accepted it.

Agreed upon flag (so far): `--conditional-directives`

The implementation is already underway (checked when committed):
- [x] VM: https://codereview.chromium.org/1380383002/
- [x] dart2js: https://codereview.chromium.org/1388523002/

Still missing: environment variables for the existing libraries:
- [x] VM: issue #24587
- [x] dart2js: issue #24586

Validate implementation by updating `pkg/http`
- [x] https://github.com/dart-lang/http/issues/25

The analysis can be implemented in different phases.
- [x] analyzer: #24579
  - [x] phase 0
  - [x] phase 1  - no longer planned.
- [x] dart2js (optional): #24580   -  no longer planned.

Once the DEP committee has accepted:
- [x] remove flag
- [x] Implement in DDC https://github.com/dart-lang/dev_compiler/issues/595
- [ ] Validate behavior for summaries and bazel https://github.com/dart-lang/sdk/issues/28247

[0] https://github.com/munificent/dep-interface-libraries/blob/master/Proposal.md

**Comment by bwilkerson** [comment]
I now have a CL that will add support in the parser and AST structure for this dep (https://codereview.chromium.org/1406253004/). However, it's a breaking change that will affect many clients of analyzer, so I am hesitant to commit it before we've released 1.13.

**Comment by mit-mit** [comment]
Now that we are starting 1.14 builds, let's go ahead and land CLs that are ready for this!

**Comment by mit-mit** [comment]
Looks like the VM changes landed :-). @floitschG, what else remains in 1.14?

**Comment by floitschG** [comment]
I would like to see the env-variables landed.
I have a patch for dart2js, but things are slightly more complicated in the VM.

**Comment by floitschG** [comment]
Now available behind a flag.
Here is the Changelog update: 1d2a46e6b4d294ca45399fb6f26d88e322a34142

**Comment by mit-mit** [comment]
Great! Can we close this bug? And maybe have a new one for removing the flags?

**Comment by mit-mit** [comment]
Clearing out 1.15 milestone as the last full push to dev has happened. If any changes are required before 1.15 is shipped, please file a merge request

**Comment by sethladd** [comment]
What is phase 0 and phase 1 mean for analyzer? cc @bwilkerson

**Comment by bwilkerson** [comment]
I'm not sure, but I think phase 0 was correctly analyzing code based on the interface library and accepting the new syntax, and phase 1 was implementing the verification of the implementation libraries.

**Comment by floitschG** [comment]
The phases are described in the DEP.

**Comment by sethladd** [comment]
https://codereview.chromium.org/1778683004/ adds support for enabling this option in the CLI version of analyzer. We might want to update the above checklist?

**Comment by floitschG** [comment]
The analyzer already supported it (through the options file). The check-mark already reflected this.

**Comment by nex3** [comment]
I've locally verified that `http` works with the new system. I won't be able to actually commit that code until this is out from under the flag, but this should no longer be considered blocked on verification.

**Comment by floitschG** [comment]
Thanks Natalie.
My personal experience with the system was also quite positive.

**Comment by mit-mit** [comment]
@floitschG can we close this issue now?

**Comment by nex3** [comment]
Why was this closed? The feature is still behind a flag, and full analysis support isn't yet implemented, which means it's not safe to use in real code. As I understand it, meta-issues should only be closed once the feature is considered completely implemented.

**Comment by kevmoo** [comment]
Agreed. We should make sure all of the checkboxes are checked – and that we're not missing check boxes.

**Comment by mit-mit** [comment]
From discussing with Florian, there are no current plans to add any more validation at this time. Are we really blocked from using this (e.g., in http) with the current support?

**Comment by nex3** [comment]
Last I heard from the language team, the official plan was explicitly *not* to do validation, and that any validation that might be added in the future would be optimistic—@floitsch @munificent @leafpetersen can you confirm? If this is the case, we should remove the checkboxes above about phase 1 analysis and dart2js analysis.

If my understanding is correct, we're no longer blocked on this to upgrade packages. However, @dgrove asked that I put progress towards this on hold for the time being to prioritize dev-cycle work.

**Comment by floitschG** [comment]
Thinking more about it, we should keep this bug open, until we publicly commit to it.
We already got a request to change the syntax, and until we announce the feature we should be able to do this kind of chances. We should not keep this bug indefinitely open, though.

Wrt updating the description: I don't like removing items in existing comments. It removes context for later comments. I would be ok with updating them, stating that these are not planned anymore. (I just closed the dart2js bug).

**Comment by nex3** [comment]
Okay, maybe something like:

- [x] analyzer: #24579
  - [x] phase 0
  - [x] phase 1 *(no longer planned)*

**Comment by eernstg** [comment]
If we have syntax for validation (that is: an explicit library signature clause) then we can easily implement phase 1 whenever we want. If we don't include that syntax then we can't ever hope to check anything, because it will break existing unchecked usage. So we don't have to send that "no longer planned" message if we want to keep this door open.

**Comment by nex3** [comment]
> If we have syntax for validation (that is: an explicit library signature clause) then we can easily implement phase 1 whenever we want

I'm not sure what you mean. We definitely have interface libraries, which could be interpreted as a "library signature clause", but the rules for what counts as a "compatible signature" are vague—and necessarily will be until an implementation of verification exists or a similarly-comprehensive spec is written.

> If we don't include that syntax then we can't ever hope to check anything, because it will break existing unchecked usage. So we don't have to send that "no longer planned" message if we want to keep this door open.

Breaking existing unchecked usage is something the language has done before, so "it would be a breaking change in practice" isn't enough of a guarantee for me that the change won't actually be made.

@mit-mit One thing I forgot to mention was that we *are* blocked on this being considered experimental and hidden behind a flag. We can't update existing packages to use this until it's enabled by default in all implementations.

**Comment by mit-mit** [comment]
@nex3, @floitschG so all that we are blocked on is removing all the flags we introduced for this in a concerted fashion? And would we do that in 1.18?

**Comment by floitschG** [comment]
Blocked on announcing it (thus committing to it). At the same time we need to update the spec.
1.18 is an ambitious target, but let's try.

**Comment by eernstg** [comment]
>> .. an explicit library signature clause ..

> .. We definitely have interface libraries, which could be interpreted as a "library signature clause", but the rules for what counts as a "compatible signature" are vague

The point is that it should be possible to _omit_ the library signature as well as including it, which means that it can't simply be an extra meaning attached to the first choice among a list of configurably imported libraries. So all we need to do is to make the interface library optional (or make it a pure default choice that has no special role during static analysis and then allow for an explicit library signature, say, "`import` uri configuration* (`as` identifier)? combinator* (`signature` uri)? `;`").

When there is no library signature there will be no checking discipline imposed on the set of configurably imported libraries, but tools may of course check any specific configuration. This yields full checking per configuration plus optimal flexibility, but no encapsulation.

Level 1 checking is trivial: Only top-level non-type declarations are allowed, and each configurably imported library must deliver top-level declarations with the same names and assignable values. All other declarations in the configurably imported libraries are hidden, just like SML signatures. This yields full encapsulation (outside configurably imported libraries the program is strictly configuration independent) and modular checking.

Without an explicit library signature we can't hope to offer such a clear and powerful choice to programmers.

**Comment by kevmoo** [comment]
Seems that DDC *won't* be ready for this for 1.19.

Also: I don't see any impl of pkg/http (as we discussed) that uses this feature.

Both are blockers to call it "done".

**Comment by nex3** [comment]
I didn't end up getting any time allocated towards this this quarter. It's worth nothing that we can't release a version of `http` that supports this until it's no longer behind a flag, though.

**Comment by mit-mit** [comment]
Let's not rush this, bumping this to 1.20

**Comment by nex3** [comment]
Heads up, it looks like dart2js support is kind of broken: #27548

**Comment by stevenroose** [comment]
Awesome! I'll try to rewrite [`package:websockets`](https://pub.dartlang.org/packages/websockets) to incorporate this next week.

**Comment by stevenroose** [comment]
How can I test this in a browser? Both dart2js and pub run test fail with
`Error: Library not supported 'dart:io'.`.

**Comment by kevmoo** [comment]
@stevenroose Hold up. There are still things that are not implemented yet – especially in the analyzer.

We recommend folks be very cautious implementing this feature – especially in heavily shared code.

**Comment by nex3** [comment]
> Hold up. There are still things that are not implemented yet – especially in the analyzer.

AFAIK Analyzer has finished all planned support for this. There certainly aren't any tasks above listed as incomplete.

> We recommend folks be very cautious implementing this feature – especially in heavily shared code.

I believe the only thing stopping this from being used, at least externally, is that it's still behind a flag in all implementations.

**Comment by stevenroose** [comment]
Yeah I just wanted to test this. @nex3 how do you get dart2js to work with
this? It doesn't accept the same flag as dart does..

**Comment by nex3** [comment]
I think dart2js actually supports this without a flag, although some of its import-resolution logic was broken prior to 1.20.

**Comment by eseidelGoogle** [comment]
@kevmoo or @mit-mit : who's the right person to drive the rest of this forward (I see it's unassigned)?  My understanding is this is blocking https://github.com/dart-lang/http/issues/1 which has come up several times for Flutter customers attempting to depend on http-using packages in pub.

**Comment by stevenroose** [comment]
I'd really want to see this land too. [package:websockets](https://github.com/stevenroose/dart-websockets/) would benefit equally from this as package:http does.

**Comment by eseidelGoogle** [comment]
I just spoke with @dgrove, I'm told this is largely blocked on analyzer support, which in turn is blocked on more important analyzer issues, and we should expect movement on this issue in early 2017.

**Comment by mit-mit** [comment]
Marking DDC done based on https://github.com/dart-lang/sdk/issues/27302#issuecomment-267076645 👍

**Comment by mit-mit** [comment]
@bwilkerson is there any remaining critical work in the analyzer?

**Comment by bwilkerson** [comment]
I honestly don't know. It's been too long since I last looked at it.

Do we have language tests? If so, is analyzer passing them?

I saw an e-mail or issue a few days ago indicating that someone thought analyzer wasn't choosing the right URI to analyze against, but I haven't had time to determine whether that's true.

**Comment by lrhn** [comment]
I've added another language test. The analyzer is passing in the sense that I had to mark it as giving the warnings that it should, and otherwise accepts the test. I'm not testing the individual warnings, but they look correct to me.

**Comment by mit-mit** [comment]
Lasse's test is https://codereview.chromium.org/2579253003/ -- looks good to me!

**Comment by mit-mit** [comment]
@bwilkerson anything else you want to have verified before I close this feature as complete!?

**Comment by bwilkerson** [comment]
Nope.

**Comment by mit-mit** [comment]
\o/ 

Closing (just in time for Christmas)!

**Comment by jodinathan** [comment]
When can we use it?
Are there any examples?

thanks for this, it is very important =]

**Comment by mit-mit** [comment]
The test above is probably the best source of information right now. We will have more docs ready for the 1.22 launch.

**Comment by stevenroose** [comment]
This is awesome! Great work to get this out right before the holidays :)

**Comment by sethladd** [comment]
Congrats!

**Comment by skybrian** [comment]
What about strong mode summaries for DDC? Will they be generated correctly?

**Comment by kevmoo** [comment]
That's the last bit we need to figure out...

**Comment by kevmoo** [comment]
Reopening and adding a tracking item around summary validation...

**Comment by sethladd** [comment]
Should we add summary validation to the checkboxes in the description for this issue?

**Comment by jodinathan** [comment]
Hi,

Is there any date we can expect this?
Using hack to check environment and ifs is getting bigger each day

**Comment by lrhn** [comment]
This feature is in Dart 2, but restricted to only be able to test against a number of platform specific definitions - currently one per `dart:` library, so you can test `dart.libraries.js` to check if JS integration is available.
This allows some build tools to ensure that there is only a fixed known number of different configurations.

**Comment by pulyaevskiy** [comment]
Just wondering if there are docs available about this feature with some more details?

**Comment by anders-sandholm** [comment]
Beyond https://github.com/munificent/dep-interface-libraries/blob/master/Proposal.md and @lrhn 's comment above, unfortunately not at this point. Not the answer I'd have liked to give, but currently this is not as well documented as we'd like. Stay tuned.

@lrhn @Sfshaza @kwalrath FYI.
